spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java

```diff
@@ -608,4 +712,162 @@ else if (destinationDirectoryToUse instanceof File) {
+		return new BufferedOutputStream(new FileOutputStream(fileToWriteTo, append), this.bufferSize);
+	}
+
+	/**
+	 * When using {@link FileExistsMode#APPEND_NO_FLUSH} you can send a message to this
+	 * method to flush any file(s) that needs it. By default, the payload must be a regular
+	 * expression ({@link String} or {@link Pattern}) that matches the absolutePath
+	 * of any in-process files. However, if a custom {@link FlushPredicate} is provided,
+	 * the payload can be of any type supported by that implementation.
+	 * @since 4.3
+	 */
+	@Override
+	public synchronized void trigger(Message<?> message) {
+		Iterator<Entry<String, FileState>> iterator = FileWritingMessageHandler.this.fileStates.entrySet().iterator();
+		while (iterator.hasNext()) {
+			Entry<String, FileState> entry = iterator.next();
+			FileState state = entry.getValue();
+			if (this.flushPredicate.shouldFlush(entry.getKey(), state.lastWrite, message)) {
+				iterator.remove();
+				state.close();
+			}
+		}
+	}
+
+	private static final class FileState {
+
+		private final BufferedWriter writer;
+
+		private final BufferedOutputStream  stream;
+
+		private volatile long lastWrite;
+
+		private FileState(BufferedWriter writer) {
+			this.writer = writer;
+			this.stream = null;
+		}
+
+		private FileState(BufferedOutputStream stream) {
+			this.writer = null;
+			this.stream = stream;
+		}
+
+		private void close() {
+			try {
+				if (this.writer != null) {
+					this.writer.close();
+				}
+				else {
+					this.stream.close();
+				}
+			}
+			catch (IOException e) {
+				;
+			}
+		}
+	}
+
+	private final class Flusher implements Runnable {
+
+		@Override
+		public void run() {
+			synchronized (FileWritingMessageHandler.this) {
+				long expired = FileWritingMessageHandler.this.flushTask == null ? Long.MAX_VALUE
+						: (System.currentTimeMillis() - FileWritingMessageHandler.this.flushInterval);
+				Iterator<Entry<String, FileState>> iterator = FileWritingMessageHandler.this.fileStates.entrySet().iterator();
+				while (iterator.hasNext()) {
+					Entry<String, FileState> entry = iterator.next();
+					FileState state = entry.getValue();
+					if (state.lastWrite < expired) {
+						iterator.remove();
+						state.close();
+						if (logger.isDebugEnabled()) {
+							logger.debug("Flushed: " + entry.getKey());
+						}
+					}
+				}
+			}
+		}
+
+	}
+
+	/**
+	 * When using {@link FileExistsMode#APPEND_NO_FLUSH} and a trigger message is invoked
+	 * an implementation of this interface is called for each file that has pending data
+	 * to flush.
+	 * @since 4.3
+	 *
+	 */
+	public interface FlushPredicate {
+
+		/**
+		 * @param fileAbsolutePath the path to the file.
+		 * @param lastWrite the time of the last write - {@link System#currentTimeMillis()}.
+		 * @param triggerMessage the trigger message received.
+		 * @return true if the file should be flushed.
+		 */
+		boolean shouldFlush(String fileAbsolutePath, long lastWrite, Message<?> triggerMessage);
```

Comment:
:+1: 
`flushOnPredicate(FlushPredicate flushPredicate) {...}`
But... that one should be already without any message argument...